### PR TITLE
xds: prefer fed state gateway definitions if they're fresher

### PR DESCRIPTION
**Pull request by boxofrad**
Fixes an issue (described in #10132) where if two DCs are WAN federated over mesh gateways, and the gateway in the non-primary DC is terminated and receives a new IP address (as is commonly the case when running them
on ephemeral compute instances) the primary DC is unable to re-establish its connection until the agent running on its own gateway is restarted.

This was happening because we always preferred gateways discovered by the `Internal.ServiceDump` RPC (which would fail because there's no way to dial the remote DC) over those discovered in the federation state, which is replicated as long as the primary DC's gateway is reachable - see the comment in `endpointsFromSnapshotMeshGateway`.

### CODE REVIEW

File: agent/xds/endpoints.go
Comment (boxofrad) [inline agent/xds/endpoints.go:162]: I'm assuming here that `len(fedStateEndpoint) * len(endpoints)` is small. Maybe this should be a map lookup instead?

File: agent/xds/endpoints.go
Comment (rboyer) [inline agent/xds/endpoints.go:173]: For a moment I thought this was a raft index apples/oranges comparison, but as long as you stick to the Service part of a CheckServiceNode entity in `FedStateGateways` that is going to be the same as the original data fetched in `GatewayGroups`

File: agent/xds/endpoints.go
Comment (rboyer) [inline agent/xds/endpoints.go:126]: How would you feel about this alternative?

1. collect the max inner `Service.RaftIndex.ModifyIndex` value from all MGW instances in `GatewayGroups`
2. do the same for `FedStateGateways` instances
3. depending upon which of the two data sources has the larger value, wholesale use the entire slice of data that came from it

The argument here is that on their own, each of these data sets come from singular RPC calls awoken from single blocking queries. They are each independently self-consistent, whichever one has more recent data by necessity would be in aggregate more correct, so there's no need to do a deep merge like this. Doing the either/or merge instead would also be easier to grok.

Comment (boxofrad) [reply agent/xds/endpoints.go]: That's much easier to reason about, thanks! 🙌🏻

File: agent/proxycfg/testing.go
Comment (dhiaayachi) [inline agent/proxycfg/testing.go:1510]: Do we have a test that validate the other alternative?  If not it would be good to have a test that verify switching from a data source to the other work properly.

**Review by dhiaayachi — APPROVED** [review]
I have a question about a possible missing test, but other then that it LGTM!!

File: agent/proxycfg/testing.go
Comment (boxofrad) [reply agent/proxycfg/testing.go:1510]: Great point! Have added ✅

**Review by rboyer — APPROVED** [review]
LGTM

**Comment by hc-github-team-consul-core** [comment]
:cherries: If backport labels were added before merging, cherry-picking will start automatically.

To retroactively trigger a backport after merging, add backport labels and re-run https://circleci.com/gh/hashicorp/consul/497440.

**Comment by hc-github-team-consul-core** [comment]
:cherries::x: Cherry pick of commit 50a1f20ff9b0257c5384a5100e8e63c86114fd76 onto `release/1.10.x` failed! [Build Log](https://circleci.com/gh/hashicorp/consul/497440)

**Pull request merged**
